packages/core/src/model.ts

```diff
@@ -0,0 +1,40 @@
+import {
+  ActionObject,
+  AssignAction,
+  Assigner,
+  PropertyAssigner,
+  assign
+} from '.';
+import { EventObject } from './types';
+
+export interface ContextModel<TC, TE extends EventObject> {
+  context: TC;
+  actions: {
+    [key: string]: ActionObject<TC, TE>;
+  };
+  assign: <TEventType extends TE['type'] = TE['type']>(
```

Comment:
Just adding this comment for completeness.

I was experimenting with some overloads:
```diff
diff --git a/packages/core/src/model.ts b/packages/core/src/model.ts
index 71684848..ab2272a8 100644
--- a/packages/core/src/model.ts
+++ b/packages/core/src/model.ts
@@ -7,17 +7,24 @@ import {
 } from '.';
 import { EventObject } from './types';
 
+interface ModelAssign<TC, TE extends EventObject> {
+  <TEventType extends TE['type'] = TE['type']>(
+    assigner:
+      | Assigner<TC, Extract<TE, { type: TEventType }>>
+      | PropertyAssigner<TC, Extract<TE, { type: TEventType }>>,
+    eventType?: TEventType
+  ): AssignAction<TC, Extract<TE, { type: TEventType }>>;
+  <SubEvent extends TE = TE>(
+    assigner: Assigner<TC, SubEvent> | PropertyAssigner<TC, SubEvent>
+  ): AssignAction<TC, SubEvent>;
+}
+
 export interface ContextModel<TC, TE extends EventObject> {
   context: TC;
   actions: {
     [key: string]: ActionObject<TC, TE>;
   };
-  assign: <TEventType extends TE['type'] = TE['type']>(
-    assigner:
-      | Assigner<TC, TE & { type: TEventType }>
-      | PropertyAssigner<TC, TE & { type: TEventType }>,
-    eventType?: TEventType
-  ) => AssignAction<TC, TE & { type: TEventType }>;
+  assign: ModelAssign<TC, TE>;
 }
```

but at the end of the day, I haven't really found out this to be useful. The only thing that this has allowed was this:
```ts
userModel.assign<UpdateNameEvent>({
  name: (ctx, ev) => ev.type
});
```
as event this:
```ts
userModel.assign({
  name: (ctx, ev: UpdateNameEvent) => ev.type
});
```
is matching the first overload (what is in this PR).

One additional observation is also that this is possible and this feels more natural:
```ts
userModel.assign<'updateName'>({
  name: (ctx, ev) => ev.type
});
```
but this doesn't provide any autocompletes like the version you have been using here:
```ts
userModel.assign({
  name: (ctx, ev) => ev.type
}, 'updateName');
```